main.go

```diff
@@ -31,10 +43,12 @@ Daemon control:
   saptune daemon [ start | status | stop ]
 Tune system according to SAP and SUSE notes:
   saptune note [ list | verify ]
-  saptune note [ apply | simulate | verify | customise | revert ] NoteID
+  saptune note [ apply | simulate | verify | customise ] NoteID
 Tune system for all notes applicable to your SAP solution:
   saptune solution [ list | verify ]
-  saptune solution [ apply | simulate | verify | revert ] SolutionName
+  saptune solution [ apply | simulate | verify ] SolutionName
+Revert all parameters tuned by the SAP notes or solutions:
+  saptune revert all
```

Comment:
I guess this will be a breaking change since the API to the command changes. Would it be possible to still support the old "saptune solution revert .." syntax, or is it only possible to revert all tunings at once now?

Reply:
Yes your right. That will break the API. And I'm not really happy with that, but see not other solution at the moment. And that was discussed with Alliance and PM and was accepted for the first step. In a second step it should be possible to go back to the 'old' behavior.
So what is the problem:
In the past saptune sets only some hard coded system parameters and some dedicated predefined values using sysconfig files. So there was never the situation, that one system parameter was defined in two or more config files with - may be - different values. And reverting a single note or a single solution does work without problems.
The problem occurs as we added the vendor/partner support for HPE. Now it may be possible that a parameter set in saptune directly will set additional by a partner file, but to another value. And now the revert of a note may result in wrong system values, as it depends on the order in which the notes are applied. The revert must be in the revers order to get the right system values back. Howard would handle that by always apply the vendor file last. But that does not really solve the problem.
And now we totally skip the hard coded values and the predefined sysconfig values. Every system parameter is set in a Note definition file. These Note definition files cover settings described in SAP Notes or vendor/partner specific settings. And now every system parameter can be defined in more than one file and set to different value depending in the needs of the Note definition file. So when you now apply a solution (bunch of Note definition files) or one or more single notes, the 'saved state' files in /var/lib/saptune/saved_state may contain different 'old' values for the same parameter depending on the order of applying the related note.
When you now revert a single note, which only has 'unified' parameter - everything is ok. The resulting system values are the same as before applying this note. But if this note contains parameters, which are applied with different values by another note later, then the resulting system parameters after this revert are wrong. And when this other note will be reverted later, then the resulting system parameters are NOT the same as before applying any note.
So at the moment only a 'revert all', which reverts every note in revers order of the applying, is possible.
To solve the problem we need to find a new way of storing 'old' system values to be able to revert every time the right parameter value. So may be we need a 'parameter' centered saved state file (as tuned is using) instead of the 'note' centered saved state file or some other magic within the code. But this is a very complex problem. And may be I'm simply lost in that problem and didn't  see an easy solution.
So let us discuss face to face in 2 weeks. May be it's a little bit easier to show the problem and to find a working solution.

Reply:
Hmm yes, I think I understand the problem. Lets discuss further, there might be another way. One possibility could be to refuse to revert an ambiguous note, and add an option to specify which order to revert such notes. This would depend on being able to scan through and find other notes that may have the same parameter being set.

Reply:
Will be solved in a later sprint. At the moment it's an ongoing discussion.